File: leetcode-cn/Python/22.括号生成.py
#
# @lc app=leetcode.cn id=22 lang=python3
#
# [22] 括号生成
#

# @lc code=start
from typing import List


class Solution:
    # def generateParenthesis(self, n: int) -> List[str]:
    #     if n == 0:
    #         return ['']

    #     ans = []
    #     for left in range(n):
    #         right = n - left - 1
    #         a_list = self.generateParenthesis(left)
    #         b_list = self.generateParenthesis(right)
    #         for a in a_list:
    #             for b in b_list:
    #                 ans.append('({}){}'.format(a, b))

    #     return ans

    def generateParenthesis(self, n: int) -> List[str]:
        ans = []

        def help(s, left, right):
            if len(s) == 2 * n:
                ans.append(''.join(s))
                return

            if left < n:
                s.append('(')
                help(s, left + 1, right)
                s.pop()

            if right < left:
                s.append(')')
                help(s, left, right + 1)
                s.pop()

        help([], 0, 0)
        return ans


# @lc code=end

File: leetcode-cn/C++/22.括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */

// @lc code=start
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    void help(vector<string>& ans, string path, int left, int right, int n)
    {
        if (left == n && right == n) {
            ans.push_back(path);
            return;
        }
        if (left < n) {
            path.push_back('(');
            help(ans, path, left + 1, right, n);
            path.pop_back();
        }

        if (right < left) {
            path.push_back(')');
            help(ans, path, left, right + 1, n);
            path.pop_back();
        }
    }

    vector<string> generateParenthesis(int n)
    {
        vector<string> ans;
        string path;
        help(ans, path, 0, 0, n);
        return ans;
    }
};
// @lc code=end
